### DISCUSSION TITLE
Hot reloading is slow

### DISCUSSION THREAD

**Discussion by rohitsathish** [zauberzeug/nicegui#2155] [Q&A]
### Question

NiceGUI hot reloading is very slow, which makes development quite frustrating. I tried from VSCode Terminal (slow hot reloading ~5 secs, certain page link loading is ~5 secs) and Windows command prompt (no hot reloading at all).

Setup - 
OS: Windows
Environment: Conda
Python version: 3.10.8
IDE: VSCode
Version: 1.4.3

**Comment by rodja** [comment]
Wow. That's bad. Anyone else who can reproduce this?

**Reply by thetableman** [reply]
Yep! Had it for months along with a lot of traceback errors from `debugpy`, it all started around the release of NiceGUI 1.3.

OS: Windows 10 & 11
IDE: VSCode
Version: 1.4.3
Python: 3.9.4

**Reply by rohitsathish** [reply]
@thetableman did rolling back to a specific version of NiceGUI or/and python work for you? I'm still struggling with this.

**Reply by thetableman** [reply]
I haven't rolled back NiceGUI to test if there's any improvement, there's no benefit to me trying this as I need the features introduced in 1.3x and now 1.4x.

It was a while ago now but I did manage to replicate the issue with a basic FastAPI app so I feel the issue (or at least in my case) is more likely related to the underlying APIs - FastAPI, Starlette, Uvicorn, etc

**Comment by falkoschindler** [comment]
@rohitsathish What Python version are you working with? I'm just asking because we've got reports (https://github.com/zauberzeug/nicegui/issues/2138) about issues with Python 3.12 on Windows.

**Reply by rohitsathish** [reply]
Its Python 3.10.8

**Reply by frankhuurman** [reply]
Is it happening with any small code snippet or with larger projects?
I've just spun up my Win10 VM and installed Python 3.10.8 to test the most basic ui.label example and it takes about 3 seconds in that VM from saving the file(in Git Bash) to seeing the updated label text in the browser which seems fair for the usual slowness of a VM.

Just to throw in my 2 cents from my own experience, the Win10 and Python 3.11.x combination seems to be the sweet spot for working on a NiceGUI app. Unfortunately Windows development ruins the developer experience every now and then.

**Reply by rohitsathish** [reply]
My app is pretty minimal, so I don't think thats the issue. 
With a Ubuntu WSL it works perfectly with quick reloads, so I guess I'll stick with that. 
But would be nice to work from Windows eventually.

**Reply by rodja** [reply]
@rohitsathish could you try/compare with a super simple hello world app? Is reloading still slow? Another worth experiment would be a plain FastAPI app (as @frankhuurman tried in https://github.com/zauberzeug/nicegui/issues/2138#issuecomment-1849059667).

**Reply by pdekeulenaer** [reply]
Same thing here. Windows 10, Python 3.12 and Nicegui 1.4.8. It basically does not reload at all. I ran some tests with Python 3.11 and there it works better (it updates every time I refresh the browser)

Can't find any error traces but happy to share info if you can tell me where to look

**Reply by rodja** [reply]
@pdekeulenaer could you try a plain FastAPI app (as @frankhuurman tried in https://github.com/zauberzeug/nicegui/issues/2138#issuecomment-1849059667)?

**Comment by maxencealluin** [comment]
Any update on this? Using nicegui 2.11.0 and python 3.10.8, the hot reload doesn't really work, it is either really long (>10s) or hangs up forever.

**Reply by falkoschindler** [reply]
Hi @rohitsathish, I'm not aware of any updates.

Is hot reload slow for _every_ app, even a short hello-world example?

Maybe you can try the following minimal FastAPI example:

main.py:
```py
#!/usr/bin/env python3
import uvicorn
from fastapi import FastAPI
from fastapi.responses import PlainTextResponse

app = FastAPI()

@app.get("/", response_class=PlainTextResponse)
def index():
    return "Hello, World!"

if __name__ == "__main__":
    uvicorn.run('main:app', host="0.0.0.0", port=8080, reload=True)
```

**Reply by sht2017** [reply]
I just tried the example. I confirm it works well with Python 3.11 venv (with the latest NiceGUI). But somehow uvicorn won't terminate by ctrl+c with Windows (encode/uvicorn#1972) and it has been a while so I presume there might be an update of uvicorn broke its ability to terminate itself properly (might need to investigate to the effect from encode/uvicorn#2289). And that may cause problems with the hot reload. Lately a similar issue I faced was using Python 3.12 with litestar a few months ago. Uvicorn caused hot reload with litestar is unlikely to work for most cases to me.

**Comment by s71m** [comment]
I solved the reload problem by simply reimporting modules. It's much faster than restarting the server.
